src/Form/Expression/ConditionContainerForm.php

```diff
@@ -24,65 +25,356 @@ class ConditionContainerForm implements ExpressionFormInterface {
+              'expression_id' => 'rules_and',
+            ]),
+          ],
+        ),
+        array(
+          '#theme' => 'menu_local_action',
+          '#link' => [
+            'title' => $this->t('Add OR'),
+            'url' => $this->getRulesUiHandler()->getUrlFromRoute('expression.add', [
+              'expression_id' => 'rules_or',
+            ]),
+          ],
+        ),
+      ),
+    );
+
+    // Table footer.
+    $form['conditions']['table']['#footer'] = [
+      array(
+        'class' => array('footer-class'),
+        'data' => array(
+          array(
+            'data' => $footer,
+            'colspan' => 2,
+          ),
+        ),
+      ),
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    $values = $form_state->getValue('table');
+    if ($values) {
+      $elements = $this->mirrorElements($values);
+      $tree = $this->buildElementTree($values, $elements, 'parent', 'id');
+      $this->saveElementTree($tree, $elements);
+      // Remove the original expressions since we've cloned them to a new tree.
+      foreach ($values as $uuid => $old_expression) {
+        $this->ruleExpression->deleteExpression($uuid);
+      }
+      $this->getRulesUiHandler()->updateComponent($this->component);
+    }
+  }
+
+  /**
+   * Build a tree of values from a flat array.
+   *
+   * @param array $values
+   *   An associative array.
+   * @param string $pidKey
+   *   The parent id key in the values.
+   * @param string $idKey
+   *   The id key in the values.
+   *
+   * @return mixed
+   *   A heirarchical tree of values.
+   */
+  private function buildElementTree($values, &$elements, $pidKey, $idKey = NULL) {
+    $grouped = array();
+    foreach ($values as $sub) {
+      $grouped[$sub[$pidKey]][] = $sub;
+    }
+
+    $fnBuilder = function ($siblings) use (&$fnBuilder, &$elements, $grouped, $idKey, $elements) {
+      foreach ($siblings as $k => $sibling) {
+        $id = $sibling[$idKey];
+        if (isset($grouped[$id])) {
+          $sibling['conditions'] = $fnBuilder($grouped[$id]);
+          foreach ($sibling['conditions'] as $condition) {
+            // Tell the child who the parent is by setting its root.
+            $child = $elements[$condition[$idKey]];
+            // Add the child to the parent object.
+            $elements[$id]->addExpressionObject($child);
+          }
+        }
+        $siblings[$k] = $sibling;
+      }
+      return $siblings;
+    };
+    $tree = $fnBuilder($grouped[0]);
+    return $tree;
+  }
+
+  /**
+   * Creates "clones" of submitted elements for use in a new condition tree.
+   *
+   * @param array $values
+   *   The array of form values.
+   *
+   * @return array|mixed
+   *   The derived condition expression elements.
+   */
+  private function mirrorElements($values) {
+    $elements = NULL;
+    $expressionManager = \Drupal::service('plugin.manager.rules_expression');
```

Comment:
I agree that dependency injection would be ideal. The reason its not implemented with DI is because the form is being called in a non-standard way (not from the container). Its getting a single argument passed into the constructor, and so we would have to go further up into the api to affect that. I wanted to stay away from making api changes as much as possible.